pgrx/src/pg_catalog/pg_proc.rs

```diff
@@ -0,0 +1,316 @@
+//LICENSE Portions Copyright 2019-2021 ZomboDB, LLC.
+//LICENSE
+//LICENSE Portions Copyright 2021-2023 Technology Concepts & Design, Inc.
+//LICENSE
+//LICENSE Portions Copyright 2023-2023 PgCentral Foundation, Inc. <contact@pgcentral.org>
+//LICENSE
+//LICENSE All rights reserved.
+//LICENSE
+//LICENSE Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+use crate::{pg_sys, FromDatum, IntoDatum};
+use std::ptr::NonNull;
+
+/// Provides a safe wrapper around a Postgres "SysCache" entry from `pg_catalog.pg_proc`.
+pub struct PgProc {
+    inner: NonNull<pg_sys::HeapTupleData>,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ProArgMode {
+    In,
+    Out,
+    InOut,
+    Variadic,
+    Table,
+}
+
+impl From<i8> for ProArgMode {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b'i' => ProArgMode::In,
+            b'o' => ProArgMode::Out,
+            b'b' => ProArgMode::InOut,
+            b'v' => ProArgMode::Variadic,
+            b't' => ProArgMode::Table,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProArgMode`: `{}`", value),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ProKind {
+    Function,
+    Procedure,
+    Aggregate,
+    Window,
+}
+
+impl From<i8> for ProKind {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b'f' => ProKind::Function,
+            b'p' => ProKind::Procedure,
+            b'a' => ProKind::Aggregate,
+            b'w' => ProKind::Window,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProKind`: `{}`", value),
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum ProVolatile {
+    Immutable,
+    Stable,
+    Volatile,
+}
+
+impl From<i8> for ProVolatile {
+    fn from(value: i8) -> Self {
+        match value as u8 {
+            b'i' => ProVolatile::Immutable,
+            b's' => ProVolatile::Stable,
+            b'v' => ProVolatile::Volatile,
+
+            // there's just no ability to move forward if given a value that we don't know about
+            _ => panic!("unrecognized `ProVolatile`: `{}`", value),
```

Comment:
nit:
```suggestion
            v => panic!("unrecognized `ProVolatile`: `{:?}`", v as char),
```
since these use ASCII it will probably be easier to debug if it's formatted as a character than a number.